Guide on multi-mode lexing (#132)
langium/langium-website: hugo/content/guides/multi-mode-lexing.md

---
title: "Multi-Mode Lexing"
weight: 400
---

Many modern programming languages such as [JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) or [C#](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated) support template literals.
They are a way to easily concatenate or interpolate string values while maintaining great code readability.
This guide will show you how to support template literals in Langium.

For this specific example, our template literal starts and ends using backticks `` ` ``  and are interupted by expressions that are wrapped in curly braces `{}`.

> So in our example, usage of template literals might look something like this:

Reviewer: Can drop this one

```js
println(`hello {name}!`);
```

Conceptually, template strings work by reading a start terminal which starts with `` ` `` and ends with `{`, 
followed by an expression and then an end terminal which is effectively just the start terminal in reverse using `}` and `` ` ``.

> Since we don't want to restrict users to only a single expression in their template literals, we also need a "middle" terminal reading from `}` to `{`.

Reviewer: Resolving as less relevant.

However, depending on the structure of your existing grammar, some of these new terminals might be in conflict with existing terminals of your language.
For example, if your language supports block statements, chaining multiple blocks together will make this issue apparent:

```js
{
    console.log('hi');
}
{
    console.log('hello');
}
```

The `} ... {` block in this example won't be parsed as separate `}` and `{` tokens, but instead as a single `TEMPLATE_LITERAL_MIDDLE` token, resulting in a parser error due to the unexpected token.
This doesn't make a lot of sense, since we aren't in the middle of a template literal at this point anyway.
However, our lexer doesn't know yet that the `TEMPLATE_LITERAL_MIDDLE` and `TEMPLATE_LITERAL_END` terminals are only allowed to show up within a `TemplateLiteral` rule.
To rectify this, we will need to make use of lexer modes. They will give us the necessary context to know whether we're inside a template literal or outside of it.
Depending on the current selected mode, we can lex different terminals. In our case, we want to exclude the `TEMPLATE_LITERAL_MIDDLE` and `TEMPLATE_LITERAL_END` terminals.

> The following implementation of a `TokenBuilder` will do the job for us. It creates two lexing modes, which are almost identical except for the `TEMPLATE_LITERAL_MIDDLE` and `TEMPLATE_LITERAL_END` terminals.

Reviewer: There should be at least another sentence devoted to what the TokenBuilder is, and another one that describes what the 2 lexing modes are in this context.

We will also need to make sure that the modes are switched based on the `TEMPLATE_LITERAL_START` and `TEMPLATE_LITERAL_END` terminals. We use `PUSH_MODE` and `POP_MODE` for this.

> ```ts

Reviewer: As a step up from this, I still feel we should split this up. But in the interest of moving this along after some time can we instead make an issue for a custom token builder guide separately?

import { DefaultTokenBuilder, isTokenTypeArray, GrammarAST } from "langium";
import { IMultiModeLexerDefinition, TokenType, TokenVocabulary } from "chevrotain";

const REGULAR_MODE  = 'regular_mode';
const TEMPLATE_MODE = 'template_mode';

export class CustomTokenBuilder extends DefaultTokenBuilder {

>     override buildTokens(grammar: GrammarAST.Grammar, options?: { caseInsensitive?: boolean }): TokenVocabulary {

Reviewer: I would break this out into a separate paragraph, explaining we need to first build up a multi-mode lexer definition that has various modes, which are pushed on by our special tokens.

Reviewer: See above.

        const tokenTypes = super.buildTokens(grammar, options);

        if(isTokenTypeArray(tokenTypes)) {
            // Regular mode just drops template literal middle & end
            const regularModeTokens = tokenTypes
                .filter(token => !['TEMPLATE_LITERAL_MIDDLE','TEMPLATE_LITERAL_END'].includes(token.name));
            // Template mode needs to exclude the '}' keyword
            const templateModeTokens = tokenTypes
                .filter(token => !['}'].includes(token.name));

            const multiModeLexerDef: IMultiModeLexerDefinition = {
                modes: {
                    [REGULAR_MODE]: regularModeTokens,
                    [TEMPLATE_MODE]: templateModeTokens
                },
                defaultMode: REGULAR_MODE
            };
            return multiModeLexerDef;
        } else {
            throw new Error('Invalid token vocabulary received from DefaultTokenBuilder!');
        }
    }

>     protected override buildKeywordToken(

Reviewer: This would make a nice second part, indicating we need cleanup our `}` token so regular mode doesn't get messed up.

Reviewer: See above.

        keyword: GrammarAST.Keyword,
        terminalTokens: TokenType[],
        caseInsensitive: boolean
    ): TokenType {
        let tokenType = super.buildKeywordToken(keyword, terminalTokens, caseInsensitive);
        
        if (tokenType.name === '}') {
            // The default } token will use [TEMPLATE_LITERAL_MIDDLE, TEMPLATE_LITERAL_END] as longer alts
            // We need to delete the LONGER_ALT, they are not valid for the regular lexer mode
            delete tokenType.LONGER_ALT;
        }
        return tokenType;
    }

>     protected override buildTerminalToken(terminal: GrammarAST.TerminalRule): TokenType {

Reviewer: Lastly, we can add this & explain how we're associating a push/pop action for start/end literals (which chevrotain needs).

Reviewer: See above.

        let tokenType = super.buildTerminalToken(terminal);

        // Update token types to enter & exit template mode
        if(tokenType.name === 'TEMPLATE_LITERAL_START') {
            tokenType.PUSH_MODE = TEMPLATE_MODE;
        } else if(tokenType.name === 'TEMPLATE_LITERAL_END') {
            tokenType.POP_MODE = true;
        }
        return tokenType;
    }
}
```

With this change in place, the parser will work as expected. There is one last issue which we need to resolve in order to get everything working perfectly.
When inspecting our AST, the `TemplateLiteral` object will contain strings with input artifacts in there; mainly `` ` ``, `{` and `}`.
These aren't actually part of the semantic value of these strings, so we should get rid of them.
We will need to create a custom `ValueConverter` and remove these artifacts:

```ts

> import { CstNode, GrammarAST, DefaultValueConverter, ValueType, convertString } from 'langium';

Reviewer: ```suggestion
import { CstNode, GrammarAST, DefaultValueConverter, ValueType } from 'langium';
```

export class CustomValueConverter extends DefaultValueConverter {

    protected override runConverter(rule: GrammarAST.AbstractRule, input: string, cstNode: CstNode): ValueType {
        if (rule.name.startsWith('TEMPLATE_LITERAL')) {

>             // 'convertString' simply removes the first and last character of the input
>             return convertString(input);

Reviewer: ```suggestion
            // simply removes the first and last character of the input
            return input.slice(1,-1);
```
We can use slice here, in case we ever change the Langium API and `convertString` is changed or is removed. One less thing to maintain in the docs.

Reviewer: Less important, will resolve in favor of grammar changes.

        } else {
            return super.runConverter(rule, input, cstNode);
        }
    }
}
```

> Of course, let's not forget to bind all of these services:

Reviewer: ```suggestion
Of course, let's not forget to bind all of these services in your **module.ts**:
```

```ts
export const CustomModule = {
    parser: {
        TokenBuilder: () => new CustomTokenBuilder(),
        ValueConverter: () => new CustomValueConverter()
    },
};

> ```

Reviewer: Would be good to add a concluding paragraph
